### DISCUSSION TITLE
Fragmentation works with two fragments, but not with more

### DISCUSSION THREAD

**Discussion by WolframStrauss** [inet-framework/inet#911] [General]
Hi,
the following test simulation works with two fragments (even though an access violation (`Exception code=0xc0000005 flags=0x0 at 0x00007FFC3A9F649E. Access violation - attempting to read data at address 0x0000000700000018`) comes up after closing the simulation window), but does not with three framents or more. It ends with: `<!> scheduleAt(): Message (inet::ClockEvent)ProgressTimer is currently scheduled, use cancelEvent() before rescheduling -- in module (inet::InterpacketGapInserter) FragmentationDefragmentation.gap (id=6), at t=0.2s, event #1` .

Any ideas what my mistake is, or how to implement it with other Inet blocks?

Thank you. Wolfram

```code
[General]
# SQLite export
outputvectormanager-class="omnetpp::envir::SqliteOutputVectorManager"
outputscalarmanager-class="omnetpp::envir::SqliteOutputScalarManager"

# Works as expected: x.2x and x.3s a fragment is transfered (x = 0, 1, 2 ...) if there are two fragments.
# But ends with: `Exception code=0xc0000005 flags=0x0 at 0x00007FFC3A9F649E. Access violation - attempting to read data at address 0x0000000700000018` 
# after closing the simulation window.

# Three fragments cause the simulation to terminate at 0.2s: `<!> scheduleAt(): Message (inet::ClockEvent)ProgressTimer is currently scheduled, use cancelEvent() before rescheduling -- in module (inet::InterpacketGapInserter) FragmentationDefragmentation.gap (id=6), at t=0.2s, event #1`
[Config FragmentationDefragmentation]
description = "Getting to know fragmentation / defragmentation with a gap inserter"
network = FragmentationDefragmentation
sim-time-limit = 5s

*.source.productionInterval = 1s
*.source.initialProductionOffset = 0.2s
*.source.packetLength = 600B
*.gap.duration = 0.1s
*.splitter.fragmenterPolicy.typename = "LengthBasedFragmenterPolicy"
*.splitter.fragmenterPolicy.maxFragmentLength = 300B
```

```code
package on_inet.simulations.topics;

import inet.protocolelement.common.InterpacketGapInserter;
import inet.protocolelement.fragmentation.FragmentNumberHeaderBasedDefragmenter;
import inet.protocolelement.fragmentation.FragmentNumberHeaderBasedFragmenter;
import inet.protocolelement.fragmentation.policy.LengthBasedFragmenterPolicy;
import inet.queueing.queue.PacketQueue;
import inet.queueing.server.InstantServer;
import inet.queueing.sink.PassivePacketSink;
import inet.queueing.source.ActivePacketSource;

// https://doc.omnetpp.org/inet/api-current/neddoc/packages.html#inet_protocolelement_fragmentation
network FragmentationDefragmentation
{
    submodules:
        source: ActivePacketSource {
            @display("p=100,100");
        }
        splitter: FragmentNumberHeaderBasedFragmenter {
            @display("p=300,100");
        }
        queue: PacketQueue {
            @display("p=500,200");
        }
        server: InstantServer {
            @display("p=500,300");
        }
        gap: InterpacketGapInserter {
            @display("p=500,100");
        }
        joiner: FragmentNumberHeaderBasedDefragmenter {
            @display("p=300,300");
        }
        sink: PassivePacketSink {
            @display("p=100,300");
            @statistic[dataAge](title="Packet delay"; source=dataAge(packetPushed); record=vector; unit=s; interpolationmode=none);
        }
    connections allowunconnected:
        source.out --> splitter.in;
        joiner.out --> sink.in;
        queue.out --> server.in;
        splitter.out --> gap.in;
        gap.out --> queue.in;
        server.out --> joiner.in;
}
```

**Comment by levy** [comment]
From a first glance this should work as you have done it.

**Reply by WolframStrauss** [reply]
Hi levy, good to hear. Thank you for your feedback. I will check the simulation on older Oment / Inet versions.

**Reply by WolframStrauss** [reply]
Version information: omnetpp-6.0.1, inet-4.5.0-3833582230

**Reply by WolframStrauss** [reply]
Comuter information: Windows 10 Enterprise, 22H2

**Reply by WolframStrauss** [reply]
# Fragmenter / defragmenter Checks
Simulation outputs for three fragments (i.e. `*.source.packetLength = 601B`), after concluding simulation. Inet versions `3.8.3` and `4.2.10` do not provide the necessary modules. Since version `4.2.10` is the last Inet version for Oment `5.7`, tests with this Oment version were not carried out.

| Omnet version | Inet version | Message (simulator output) | Console output |
| - | - | - | - |
| omnetpp-6.0.1 | inet-4.5.2-4e9a4990d0 | <!> scheduleAt(): Message (inet::ClockEvent)ProgressTimer is currently scheduled, use cancelEvent() before rescheduling -- in module (inet::InterpacketGapInserter) FragmentationDefragmentation.gap (id=6), at t=0.2s, event #1 | undisposed object: (inet::Packet) FragmentationDefragmentation.splitter.source-0 -- check module destructor / undisposed object: (inet::Packet) FragmentationDefragmentation.gap.source-0-frag1 -- check module destructor |
| omnetpp-6.0.1 | inet-4.5.0-3833582230 | <!> scheduleAt(): Message (inet::ClockEvent)ProgressTimer is currently scheduled, use cancelEvent() before rescheduling -- in module (inet::InterpacketGapInserter) FragmentationDefragmentation.gap (id=6), at t=0.2s, event #1 | undisposed object: (inet::Packet) FragmentationDefragmentation.splitter.source-0 -- check module destructor / undisposed object: (inet::Packet) FragmentationDefragmentation.gap.source-0-frag1 -- check module destructor |
| omnetpp-6.0.1 | inet-4.4.1-302861f35c | <!> scheduleAt(): Message (inet::ClockEvent)ProgressTimer is currently scheduled, use cancelEvent() before rescheduling -- in module (inet::InterpacketGapInserter) FragmentationDefragmentation.gap (id=6), at t=0.2s, event #1 | undisposed object: (inet::Packet) FragmentationDefragmentation.splitter.source-0 -- check module destructor / undisposed object: (inet::Packet) FragmentationDefragmentation.gap.source-0-frag1 -- check module destructor |
| omnetpp-6.0.1 | inet-4.4.0-ceafc917d5 | <!> scheduleAt(): Message (inet::ClockEvent)ProgressTimer is currently scheduled, use cancelEvent() before rescheduling -- in module (inet::InterpacketGapInserter) FragmentationDefragmentation.gap (id=6), at t=0.2s, event #1 | undisposed object: (inet::Packet) FragmentationDefragmentation.splitter.source-0 -- check module destructor / undisposed object: (inet::Packet) FragmentationDefragmentation.gap.source-0-frag1 -- check module destructor |
| omnetpp-6.0.1 | inet-4.3.9-faf9e8cea4 | <!> scheduleAt(): Message (omnetpp::cMessage)ProgressTimer is currently scheduled, use cancelEvent() before rescheduling -- in module (inet::InterpacketGapInserter) FragmentationDefragmentation.gap (id=6), at t=0.1s, event #1 | undisposed object: (inet::Packet) FragmentationDefragmentation.splitter.source-0 -- check module destructor / undisposed object: (inet::Packet) FragmentationDefragmentation.gap.source-0-frag2 -- check module destructor / undisposed object: (inet::Packet) FragmentationDefragmentation.gap.source-0-frag1 -- check module destructor |

**Comment by avarga** [comment]
This table looks nice, but it isn't very useful. Debugging the code and trying to understand what's going on would point much more towards the solution.

**Reply by WolframStrauss** [reply]
Well, it was my hope to find a quick fix for me and maybe for somebody else. Debugging is next.

**Comment by levy** [comment]
There are two issues here. The first is the address violation which is clearly a bug. It's fixed in the master branch now.

The second issue is related to the 'ProgressTimer is currently scheduled' error message. I don't know what exactly you are trying to achieve here, but the InterpacketGapInserter after the FragmentNumberHeaderBasedFragmenter will not work. The reason is that the InterpacketGapInserter creates backpressure towards its input during the IPG period. Unfortunately, the FragmentNumberHeaderBasedFragmenter sends all fragments at once, and it can't really do anything else.

You could modify your network like this: source -> splitter -> queue -> server -> gap -> joiner-> sink

**Reply by WolframStrauss** [reply]
Thank you very much.
